### PR TITLE
Migrate to C++ Persistence (1/2)

### PR DESCRIPTION
**Pull request by wilhuff**
This migrates existing Persistence components to actually use the C++ Persistence interface, along with other usages in source. Tests are coming in part 2/2.

### CODE REVIEW

**Comment by wilhuff** [comment]
Note that in addition to the diffbase on #3755, this PR only builds with CMake. The Xcode build will work once the tests land in #3757.

Consequently the actual merge target for this PR is a persistence-master branch that I'll merge once all dependent PRs have also been reviewed.

File: Firestore/core/src/firebase/firestore/local/leveldb_key.h [hunk 1]
```diff
@@ -142,7 +143,7 @@ class LevelDbMutationKey {
  private:
   // Deliberately uninitialized: will be assigned in Decode
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/leveldb_key.h]: Looks like the comment is now stale? (here and below)

File: Firestore/core/src/firebase/firestore/local/leveldb_key.h [hunk 2]
```diff
@@ -142,7 +143,7 @@ class LevelDbMutationKey {
...
   std::string user_id_;
-  model::BatchId batch_id_;
+  model::BatchId batch_id_ = model::kBatchIdUnknown;
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/leveldb_key.h]: What's the reason for the changes in this file?

File: Firestore/core/src/firebase/firestore/core/firestore_client.h [hunk 3]
```diff
@@ -55,11 +54,19 @@ OBJC_CLASS(FSTSyncEngine);
...
+namespace local {
+
+class Persistence;
+class LruDelegate;
+
+}  // namespace local
 
 namespace remote {
+
 // Forward declaration to make CMAKE build pass as it does not support
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/firestore_client.h]: Looks like this comment will become stale pretty soon.

File: Firestore/core/src/firebase/firestore/local/leveldb_query_cache.mm [hunk 4]
```diff
@@ -46,7 +47,7 @@
 FSTPBTargetGlobal* LevelDbQueryCache::ReadMetadata(leveldb::DB* db) {
   std::string key = LevelDbTargetGlobalKey::Key();
   std::string value;
-  Status status = db->Get([FSTLevelDB standardReadOptions], key, &value);
+  Status status = db->Get(StandardReadOptions(), key, &value);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/local/leveldb_query_cache.mm]: Question: is this a free function now?

File: Firestore/core/src/firebase/firestore/core/firestore_client.mm [hunk 5]
```diff
@@ -159,39 +153,39 @@ new FirestoreClient(database_info, std::move(credentials_provider),
...
+
+    auto created = LevelDbPersistence::Create(
+        std::move(dir), serializer,
+        LruParams::WithCacheSize(settings.cache_size_bytes()));
+    if (!created.ok()) {
       // If leveldb fails to start then just throw up our hands: the error is
       // unrecoverable. There's nothing an end-user can do and nearly all
       // failures indicate the developer is doing something grossly wrong so we
       // should stop them cold in their tracks with a failure they can't ignore.
       [NSException
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/core/firestore_client.mm]: Seems like something we'll have to hide behind an abstraction layer (not in this PR)?

File: Firestore/core/src/firebase/firestore/core/firestore_client.h (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/core/firestore_client.h]: The comment is stale, but there's no harm in forward declaring (and indeed some benefit, in reduced build times).

Removed.

File: Firestore/core/src/firebase/firestore/core/firestore_client.mm (hunk 5 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/core/firestore_client.mm]: Absolutely. No todo required because of the overarching work of getting this file to end in .cc.

File: Firestore/core/src/firebase/firestore/local/leveldb_key.h (hunk 1 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/leveldb_key.h]: Agreed.

File: Firestore/core/src/firebase/firestore/local/leveldb_key.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/leveldb_key.h]: I caught an uninitialized field elsewhere, searched for other possibly related fields, found these and and decided that the deliberate uninitialized fields in here were just a liability. Apologies for failing to clean up the comments.

File: Firestore/core/src/firebase/firestore/local/leveldb_query_cache.mm (hunk 4 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/local/leveldb_query_cache.mm]: Yes, though in the `firebase::firestore::local` namespace, so unlikely to collide with anything else.
